url063b9533f6_karlney_tetris_ai_tgz-pJ8-karlney_tetris_cmd_TetrisGameRunnerJ8/src/karlney/tetris/ai/AIPlayer.java

```diff
@@ -0,0 +1,155 @@
+    }
+
+    public PiecePlacer getPlacer() {
+        return placer;
+    }
+
+    public void setPlacer(PiecePlacer placer) {
+        this.placer = placer;
+    }
+
+    public BoardEvaluator getEvaluator() {
+        return evaluator;
+    }
+
+    public void setEvaluator(BoardEvaluator evaluator) {
+        this.evaluator = evaluator;
+    }
+
+    public PathFinder getPathFinder() {
+        return pathFinder;
+    }
+
+    public void setPathFinder(PathFinder pathFinder) {
+        this.pathFinder = pathFinder;
+    }
+
+
+
+    public synchronized void processInput(PlayerInput input){
+        super.processInput(input);
+        if (input == PlayerInput.INSTANT_MOVE){
+            currentPiece = destinationPiece;
+            clearDestination();
+            try {
+                commitCurrentPieceToBoard();
+            } catch (UnableToPlacePieceException e) {
+                log.info(e.getMessage());
+                stop();
+            }
+        }
+    }
+
+
+    public void clearDestination(){
+        destinationPiece = null;
+    }
+
+    public void setDestination(Piece destinationPiece) {
+        this.destinationPiece = destinationPiece;
+    }
+
+    public Piece getDestination() {
+        return destinationPiece;
+    }
+
+    public boolean hasDestination() {
+        return destinationPiece!=null;
+    }
+
+
+    /**
+     * This class executes the AI action loop
+     *
+     */
+    public class AIThread implements Runnable {
+
+        private Thread t = new Thread(this);
+
+        public void run(){
+
+            try {
+                while (isRunning()){
+
+                    //If destination is unknown then find it
+                    if (!hasDestination()){
+                        //Find the best placement of the piece according the placer - given the board, the current and the next piece
+                        Piece p = placer.bestPlacement(getEvaluator(), getBoard(), getCurrentPiece(), getNextPiece());
+                        //Set the current destination for the player
+                        setDestination(p);
+
+                        if (getCurrentPiece().getType()== PieceType.I){
+                            getCurrentPiece().getType();
+                        }
+                    }
+
+                    //Calculate the path (do this in every iteration to account for 'slow AI reaction') - the piece may have moved down since last time
+                    List<PlayerInput> path = pathFinder.findPath(getBoard(),getCurrentPiece(),getDestination());
+
+                    //Sleep to avoid AI 'cheating' of doing things too quickly
+                    if (moveDelay>0){
+                        Thread.sleep(moveDelay);
+                    }
+
+                    //Execute the first action in the path
+                    if (path.size()>0){
+                        processInput(path.get(0));
+                    }
+                }
+            }catch (InterruptedException e){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)